benchmarks/Lua/som.lua

```diff
@@ -0,0 +1,631 @@
+    local obj = {
+        items = Vector.new(size or INITIAL_SIZE)
+    }
+    return setmetatable(obj, {__index = Set})
+end
+
+function Set:size ()
+    return self.items:size()
+end
+
+function Set:each ()
+    return self.items:each()
+end
+
+function Set:has_some (fn)
+    return self.items:has_some(fn)
+end
+
+function Set:get_one (fn)
+    return self.items:get_one(fn)
+end
+
+function Set:add (obj)
+    if not self:contains(obj) then
+        self.items:append(obj)
+    end
+end
+
+function Set:remove_all ()
+    self.items:remove_all()
+end
+
+function Set:collect (fn)
+    local coll = Vector.new()
+    for e in self:each() do
+        coll:append(fn(e))
+    end
+    return coll
+end
+
+function Set:contains (obj)
+    return self:has_some(function (it) return it == obj end)
+end
+
+end -- class Set
+
+local IdentitySet = {_CLASS = 'IdentitySet'} do
+setmetatable(IdentitySet, {__index = Set})
+
+function IdentitySet.new (size)
+    local obj = Set.new(size)
+    return setmetatable(obj, {__index = IdentitySet})
+end
+
+function IdentitySet:contains (obj)
+    return self:has_some(function (it) return it == obj end)
+end
+
+end -- class IdentitySet
+
+local Entry = {_CLASS = 'Entry'} do
+
+function Entry.new (hash, key, value, next)
+    local obj = {
+        hash  = hash,
+        key   = key,
+        value = value,
+        next  = next,
+    }
+    return setmetatable(obj, {__index = Entry})
+end
+
+function Entry:match (hash, key)
+    return self.hash == hash and self.key == key
+end
+
+end -- class Entry
+
+local Dictionary = {_CLASS = 'Dictionary'} do
+
+local INITIAL_CAPACITY = 16
+local floor = math.floor
+
+function Dictionary.new (size)
+    local obj = {
+        buckets = {n = size or INITIAL_CAPACITY},
+        size    = 0,
+    }
+    return setmetatable(obj, {__index = Dictionary})
+end
+
+--[[
+    The module 'bit' is available with:
+      * LuaJIT
+      * LuaBitOp extension which is available for:
+          * Lua 5.1
+          * Lua 5.2
+    The module 'bit32' is available with:
+      * Lua 5.2
+      * Lua 5.3 when compiled with LUA_COMPAT_5_2
+    The bitwise operators are added to Lua 5.3 as new lexemes (there causes
+    lexical error in older version)
+--]]
+if _VERSION < 'Lua 5.3' then
+    local bit = bit32 or require'bit'
+    local bxor = bit.bxor
+    local rshift = bit.rshift
+
+    function Dictionary:hash (key)
+        if not key then
+            return 0
+        end
+        local hash = key:custom_hash()
+        return bxor(hash, rshift(hash, 16))
+    end
+else
+    Dictionary.hash = assert(load[[
+    --  Lua 5.3 variant with bitwise operators
+    return function (self, key)
+        if not key then
+            return 0
+        end
+        local hash = key:custom_hash()
+        return hash ~ (hash >> 16)
+    end
+]])()
+end
+
+function Dictionary:is_empty ()
+    return self.size == 0
+end
+
+function Dictionary:get_bucket_idx (hash)
+    return (hash % self.buckets.n) + 1
```

Comment:
refactored, see https://github.com/smarr/are-we-fast-yet/pull/30/commits/a981c7f062ecdbac641824cc9f421341a9b808f7